elfloader-tool/src/drivers/driver.c

```diff
@@ -53,3 +53,35 @@ int initialise_devices(void)
 
     return 0;
 }
+
+static int init_device_non_boot(struct elfloader_device *dev)
+{
+    struct elfloader_driver **drvp = __start__driver_list;
+
+    while (drvp < __stop__driver_list) {
+        struct elfloader_driver *drv = *drvp;
+        int ret = table_has_match(dev->compat, drv->match_table);
+        if (ret >= 0 && drv->init_non_boot) {
+            dev->drv = drv;
+            drv->init_non_boot(dev, drv->match_table[ret].match_data);
+        }
+        drvp++;
+    }
+
+    return 0;
```

Comment:
IF any of the setup functions return non-zero, then the loop is exited early which means that the caller needs to distinguish that this has happened. I added checks at all the callsites to handle the error cases.

Reply:
But there's nothing we can do, all we can hope is that the UART is setup first and print some debug message out. So can as well keep going instead of breaking the loop.

Reply:
There's abort()